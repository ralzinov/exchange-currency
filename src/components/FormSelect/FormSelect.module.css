.invalid > [class$="-control"],
.invalid > [class$="-control"]:hover {
    border-color: #ff6b6b;
    box-shadow: 0 0 0 1px #ff6b6b;
}
