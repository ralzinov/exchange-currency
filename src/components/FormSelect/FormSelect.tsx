import * as React from 'react';
import classNames from 'classnames';
import {WrappedFieldProps} from 'redux-form';
import Select, {ValueType} from 'react-select';
import styles from './FormSelect.module.css';
import {ValidationMessage} from '../ValidationMessage';

export interface IFormSelectOption<T = unknown> {
    label: string;
    value: T;
}

interface IFormSelectProps {
    options: IFormSelectOption[];
    className?: string;
    placeholder?: string;
}

export const FormSelect: React.FC<WrappedFieldProps & IFormSelectProps> = (props) => {
    const {error, touched, invalid} = props.meta;
    const onSelect = (option: ValueType<IFormSelectOption>) => {
        props.input.onChange((option as IFormSelectOption).value);
    };

    const className = classNames(
        invalid && touched && styles.invalid,
        props.className,
        styles.host
    );

    return (
        <>
            <Select
                className={className}
                placeholder={props.placeholder || ''}
                options={props.options}
                onChange={onSelect}
                onFocus={props.input.onFocus}
                onBlur={props.input.onBlur}
            />
            <ValidationMessage>{touched ? error : ''}</ValidationMessage>
        </>
    );
};
