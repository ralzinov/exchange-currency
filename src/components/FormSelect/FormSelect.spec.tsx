import React from 'react';
import ReactDOM from 'react-dom';
import {FormSelect, IFormSelectProps} from './FormSelect';
import {create} from 'react-test-renderer';
import {WrappedFieldProps} from 'redux-form';
import {Simulate} from 'react-dom/test-utils';

jest.mock("react-select", () => ({ options, value, onChange, }: any) => {
    const React = require('react');
    function handleChange(event: any) {
        const option = options.find((o: any) => o.value === event.currentTarget.value);
        onChange(option);
    }
    return (
        <select data-testid="select" value={value} onChange={handleChange}>
            {options.map(({ label, value }: any) => (
                <option key={value} value={value}>
                    {label}
                </option>
            ))}
        </select>
    );
});

it('renders without crashing', () => {
    const metaProps: PartialDeep<IFormSelectProps> = {
        meta: {
            error: 'Error message',
            touched: true,
            invalid: true,
            dirty: true
        },
        input: {},
        options: []
    };

    const div = document.createElement('div');
    ReactDOM.render(<FormSelect {...metaProps as IFormSelectProps} />, div);
    ReactDOM.unmountComponentAtNode(div);
});

it('should render with props in invalid state', () => {
    const props: PartialDeep<WrappedFieldProps & IFormSelectProps> = {
        placeholder: 'Some placeholder',
        className: 'input',
        meta: {
            error: 'Error message',
            touched: true,
            invalid: true,
            dirty: true
        },
        input: {},
        options: []
    };
    const component = create(
        <FormSelect {...props as (WrappedFieldProps & IFormSelectProps)}/>
    );
    expect(component.toJSON()).toMatchSnapshot();
});

it('should not render error state if untouched and pristine', () => {
    const props: PartialDeep<WrappedFieldProps & IFormSelectProps> = {
        meta: {
            error: 'Error message',
            touched: false,
            invalid: true,
            dirty: false
        },
        input: {},
        options: []
    };
    const component = create(
        <FormSelect {...props as (WrappedFieldProps & IFormSelectProps)}/>
    );
    expect(component.toJSON()).toMatchSnapshot();
});

it('should call onChange event handler', () => {
    const spyOnChange = jest.fn();
    const metaProps: PartialDeep<WrappedFieldProps & IFormSelectProps> = {
        meta: {},
        input: {
            onChange: spyOnChange
        },
        options: [{
            value: 'A',
            label: 'A',
        }, {
            value: 'B',
            label: 'B',
        }]
    };

    const div = document.createElement('div');
    ReactDOM.render(<FormSelect {...metaProps as IFormSelectProps} />, div);

    const selectEl = div.querySelector('select') as HTMLSelectElement;
    selectEl.value = 'B';
    Simulate.change(selectEl);

    expect(spyOnChange).toBeCalledWith({
        value: 'B',
        label: 'B',
    });

    ReactDOM.unmountComponentAtNode(div);
});
