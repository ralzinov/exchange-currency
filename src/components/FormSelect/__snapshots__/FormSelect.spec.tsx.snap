// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not render error state if untouched and pristine 1`] = `
Array [
  <select
    data-testid="select"
    onChange={[Function]}
  />,
  <div
    className="host"
  >
    
  </div>,
]
`;

exports[`should render with props in invalid state 1`] = `
Array [
  <select
    data-testid="select"
    onChange={[Function]}
  />,
  <div
    className="host"
  >
    Error message
  </div>,
]
`;
